compiler/rustc_codegen_llvm/src/gotoc/rvalue.rs

```diff
@@ -656,16 +656,32 @@ impl<'tcx> GotocCtx<'tcx> {
                 let src_goto_expr = self.codegen_operand(o);
                 let src_mir_type = self.operand_ty(o);
                 let dst_mir_type = t;
-                let dst_goto_expr = self.cast_sized_expr_to_unsized_expr(
-                    src_goto_expr.clone(),
-                    src_mir_type,
-                    dst_mir_type,
-                );
-                dst_goto_expr.unwrap_or(src_goto_expr)
+                self.cast_to_unsized_expr(src_goto_expr.clone(), src_mir_type, dst_mir_type)
+                    .unwrap_or(src_goto_expr)
             }
         }
     }
 
+    fn cast_to_unsized_expr(
+        &mut self,
+        src_goto_expr: Expr,
+        src_mir_type: Ty<'tcx>,
+        dst_mir_type: Ty<'tcx>,
+    ) -> Option<Expr> {
+        // Check if the cast is from a vtable fat pointer to another
+        // vtable fat pointer (which can happen with auto trait fat pointers)
+        if is_dyn_trait_fat_pointer(src_mir_type) {
+            self.cast_unsized_dyn_trait_to_unsized_dyn_trait(
+                src_goto_expr.clone(),
+                src_mir_type,
+                dst_mir_type,
+            )
+        } else {
```

Comment:
If we can add a check that the src isn't a different fat pointer type here, I think we're good to go

Reply:
Added!